### DISCUSSION TITLE
Feedback not running properly in Cognition.run

### DISCUSSION THREAD

**Discussion by dusannikolic92** [jspsych/jsPsych#2204] [Q&A] [answered]
Hello,

I am Dusan Nikolic, and I am writing a code for an AX task for my experiment. 

I am seeing some problems with Cognition.run. It won't run my code well, and the code runs fine on a server. 
The mistake Cognition seems to be making is in the feedback, that is, it always displays "Incorrect" when it should display "Correct". I tried the same code on a server, and it works, but I want to run this through Cognition to collect the data.
The code is simple - the one that you wrote in the workshop "Moving Research Online": 

```

 var probe = {
            type: 'html-keyboard-response',
            stimulus: function() {
                return '<span style="font-size:40px;">' + jsPsych.timelineVariable('probe', true) + '</span>';
            },
            choices: ['f','j'],
            post_trial_gap: 500,
            data: {
                target: jsPsych.timelineVariable('target'),
                probe: jsPsych.timelineVariable('probe'),
                condition: jsPsych.timelineVariable('condition'),
                correct_response: jsPsych.timelineVariable('correct_response'),
                trial_part: 'probe',
                task_part: jsPsych.timelineVariable('task_part')
            },
          
on_finish: function(data) {
                var acc = false;
                if (data.correct_response == jsPsych.pluginAPI.convertKeyCodeToKeyCharacter(data.key_press)) {
                    acc = true;
                }
                data.accuracy = acc;
            }
        };

var feedback = {
            type: 'html-keyboard-response',
            stimulus: function() {
                var feedback_text = '<span style="color:red;font-size:30px;">Incorrect.</span>';
                var last_resp_acc = jsPsych.data.getLastTrialData().values()[0].accuracy;
                if (last_resp_acc == true) {
                    feedback_text = '<span style="color:green;font-size:30px;">Correct!</span>'
                }
                return feedback_text;
            },
            choices: jsPsych.NO_KEYS,
            trial_duration: 2000
        };

```
I will tag in this post, the Cognition.run creator, @javidalpe, if he could help.
Thank you very much for the help and this lovely discussion forum!

**Comment by becky-gilbert** [comment, marked as answer]
Hi @dusannikolic92, are you seeing any errors in the JavaScript console when you run your task on cognition.run? Also, what version of jsPsych did you select on cognition.run? I ask because I noticed that this line:
```js
if (data.correct_response == jsPsych.pluginAPI.convertKeyCodeToKeyCharacter(data.key_press)) {
```
won't work in v6.3+ because the `key_press` property in the trial data was changed to `response`. In v6.3 we also switched to using key characters (e.g. 'f', 'j') in the data, rather than numeric key codes. So instead of using the `jsPsych.pluginAPI.convertKeyCodeToKeyCharacter` function, you would use [`jsPsych.pluginAPI.compareKeys`](https://www.jspsych.org/7.0/reference/jspsych-pluginAPI/#jspsychpluginapicomparekeys) to check the response accuracy. So altogether, the line above would change to this in jsPsych v6.3+:
```js
if (jsPsych.pluginAPI.compareKeys(data.correct_response, data.response)) {
```
If you selected jsPsych v6.3+ on cognition.run but don't want to update your code, you could just try switching to an jsPsych earlier release to see if that fixes that problem.

Another thing you can try is adding some `console.log`s in the `on_finish` function for the `probe` trial and in the `stimulus` function for the `feedback` trial, to check the values of variables such as `data` and `last_response_accuracy`.

I hope this helps. Let us know if you're able to solve the problem or if you need further guidance.

**Comment by dusannikolic92** [comment]
Hello @becky-gilbert. 

Thank you so much! This works quite well. 
In the console, I saw the warning about using the plugin convertKeyCodeToKeyCharacter, but I did not know what to replace it with. 
Also, I tried out all the Cognition versions, but the code did not function properly in any.

Anyway, it works now, thank you very much.

**Reply by becky-gilbert** [reply]
Ah, sorry about that. We should improve the console warning. Glad to hear it's working now!
